File: webUI/lib/notifications.ts
import { apiFetch } from "@/lib/api";

export interface PushConfig {
  web_push_enabled: boolean;
  web_push_public_key: string | null;
  expo_push_enabled: boolean;
}

export interface NotificationPreference {
  booking_start_push_enabled: boolean;
  booking_end_push_enabled: boolean;
}

export interface AppNotification {
  public_id: string;
  audience: string;
  reminder_type: string;
  title: string;
  body: string;
  link_url: string | null;
  is_read: boolean;
  read_at: string | null;
  push_status: string;
  push_error: string | null;
  created_at: string | null;
}

export interface NotificationList {
  results: AppNotification[];
  total: number;
  unread_count: number;
  page: number;
  page_size: number;
}

export function getPushConfig(token?: string) {
  return apiFetch<PushConfig>("/api/push/config", { method: "GET" }, token);
}

export function listNotifications(token?: string, unreadOnly = false) {
  const query = unreadOnly ? "?unread_only=true" : "";
  return apiFetch<NotificationList>(`/api/notifications${query}`, { method: "GET" }, token);
}

export function markNotificationRead(publicId: string, token?: string) {
  return apiFetch<{ status: string }>(`/api/notifications/${publicId}/read`, { method: "PATCH" }, token);
}

export function getNotificationPreferences(token?: string) {
  return apiFetch<NotificationPreference>("/api/notifications/preferences", { method: "GET" }, token);
}

export function updateNotificationPreferences(payload: Partial<NotificationPreference>, token?: string) {
  return apiFetch<NotificationPreference>(
    "/api/notifications/preferences",
    { method: "PATCH", body: JSON.stringify(payload) },
    token
  );
}

function urlBase64ToUint8Array(base64String: string) {
  const padding = "=".repeat((4 - (base64String.length % 4)) % 4);
  const base64 = (base64String + padding).replace(/-/g, "+").replace(/_/g, "/");
  const rawData = window.atob(base64);
  return Uint8Array.from([...rawData].map((char) => char.charCodeAt(0)));
}

export async function registerBrowserPush(token?: string): Promise<{ status: string; message: string }> {
  if (typeof window === "undefined" || typeof navigator === "undefined") {
    return { status: "unsupported", message: "Browser push is not available here." };
  }
  if (!("serviceWorker" in navigator) || !("PushManager" in window) || !("Notification" in window)) {
    return { status: "unsupported", message: "This browser does not support push notifications." };
  }
  const config = await getPushConfig(token);
  if (!config.web_push_enabled || !config.web_push_public_key) {
    return { status: "disabled", message: "Push notifications are not configured for this environment." };
  }
  const permission = await Notification.requestPermission();
  if (permission !== "granted") {
    return { status: permission, message: "Browser notification permission is not enabled." };
  }
  const registration = await navigator.serviceWorker.register("/push-worker.js");
  const existing = await registration.pushManager.getSubscription();
  const subscription =
    existing ??
    (await registration.pushManager.subscribe({
      userVisibleOnly: true,
      applicationServerKey: urlBase64ToUint8Array(config.web_push_public_key),
    }));
  await apiFetch(
    "/api/push-subscriptions",
    {
      method: "POST",
      body: JSON.stringify({
        provider: "web",
        subscription: subscription.toJSON(),
        user_agent: navigator.userAgent,
        device_name: "Web browser",
      }),
    },
    token
  );
  return { status: "granted", message: "Booking reminders are enabled for this browser." };
}

File: backend/app/schemas/notifications.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class PushConfigOut(BaseModel):
    web_push_enabled: bool
    web_push_public_key: str | None = None
    expo_push_enabled: bool = True


class PushSubscriptionCreate(BaseModel):
    provider: Literal["web", "expo"]
    subscription: dict[str, Any] | None = None
    token: str | None = None
    user_agent: str | None = Field(default=None, max_length=512)
    device_name: str | None = Field(default=None, max_length=255)

    @model_validator(mode="after")
    def validate_payload(self):
        if self.provider == "web":
            endpoint = (self.subscription or {}).get("endpoint")
            if not isinstance(endpoint, str) or not endpoint:
                raise ValueError("Web push subscription endpoint is required")
        if self.provider == "expo":
            if not self.token:
                raise ValueError("Expo push token is required")
        return self


class PushSubscriptionOut(BaseModel):
    public_id: str
    provider: str
    is_active: bool
    device_name: str | None = None
    last_used_at: datetime | None = None
    created_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class UserNotificationOut(BaseModel):
    public_id: str
    audience: str
    reminder_type: str
    title: str
    body: str
    link_url: str | None = None
    is_read: bool
    read_at: datetime | None = None
    push_status: str
    push_error: str | None = None
    created_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class NotificationListOut(BaseModel):
    results: list[UserNotificationOut]
    total: int
    unread_count: int
    page: int
    page_size: int


class NotificationPreferenceOut(BaseModel):
    booking_start_push_enabled: bool
    booking_end_push_enabled: bool

    model_config = ConfigDict(from_attributes=True)


class NotificationPreferenceUpdate(BaseModel):
    booking_start_push_enabled: bool | None = None
    booking_end_push_enabled: bool | None = None
